cookbook/rule/add_custom_action_rule_engine.rst

```diff
@@ -0,0 +1,429 @@
+How to add a custom action in rule engine
+=========================================
+
+Quick Overview
+--------------
+
+**This cookbook is about a feature only provided in the Enterprise Edition.**
+
+This cookbook assumes that you already created a new bundle to add your custom rule. Let's assume its namespace is `Acme\\CustomBundle`.
+
+Create a custom action
+----------------------
+
+In this cookbook we are going to see how to add a custom action in the rule engine that allow to concatenate attribute value following a pattern
+
+First let's see how to create the action. You need to create an ActionApplier object that will contain the logic:
+
+.. code-block:: php
+
+    #/src/Acme/Bundle/CustomBundle/ActionApplier/PatternActionApplier.php
+    <?php
+
+    namespace Acme\Bundle\CustomBundle\ActionApplier;
+
+    use Acme\Bundle\CustomBundle\Model\ProductPatternActionInterface;
+    use Akeneo\Bundle\RuleEngineBundle\Model\ActionInterface;
+    use Akeneo\Component\RuleEngine\ActionApplier\ActionApplierInterface;
+    use Akeneo\Component\StorageUtils\Updater\PropertySetterInterface;
+
+    class PatternActionApplier implements ActionApplierInterface
+    {
+        /** @var PropertySetterInterface */
+        protected $propertySetter;
+
+        /**
+         * @param PropertySetterInterface $propertySetter
+         */
+        public function __construct(PropertySetterInterface $propertySetter)
+        {
+            $this->propertySetter = $propertySetter;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function applyAction(ActionInterface $action, array $products = [])
+        {
+            $attributes = $action->getAttributes();
+            $pattern    = $action->getPattern();
+
+            foreach ($products as $product) {
+                $result = $pattern;
+
+                foreach ($attributes as $attributeCode) {
+                    $value = $product->getValue($attributeCode);
+
+                    $content = null === $value ? '' : (string) $value;
+                    $result = str_replace('%%' . $attributeCode . '%%', $content, $result);
+                }
+
+                $this->propertySetter->setData(
+                    $product,
+                    $action->getField(),
+                    $result,
+                    $action->getOptions()
+                );
+            }
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function supports(ActionInterface $action)
+        {
+            return $action instanceof ProductPatternActionInterface;
+        }
+    }
+
+Then we need to create the object that will handle the data, and first we will create the interface for that object:
```

Comment:
To me, its overkill to do the interface in the cookbook, we should focus the essential.